framework/src/modules/interoperability/mainchain/commands/submit_mainchain_cross_chain_update.ts

```diff
@@ -237,19 +184,122 @@ export class SubmitMainchainCrossChainUpdateCommand extends BaseCrossChainUpdate
 		} catch (error) {
 			context.eventQueue.restoreSnapshot(eventSnapshotID);
 			context.stateStore.restoreSnapshot(stateSnapshotID);
+
 			logger.info(
 				{ err: error as Error, moduleName: ccm.module, commandName: ccm.crossChainCommand },
 				'Fail to execute beforeCrossChainMessageForwarding.',
 			);
+
 			await this.internalMethod.terminateChainInternal(context, ccm.sendingChainID);
+
 			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
 				code: CCMProcessedCode.INVALID_CCM_BEFORE_CCC_FORWARDING_EXCEPTION,
 				result: CCMProcessedResult.DISCARDED,
 				ccm,
 			});
+
+			return false;
+		}
+
+		return true;
+	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#forward
+	private async _forward(context: CrossChainMessageContext): Promise<void> {
+		const { ccm } = context;
+		const { ccmID, encodedCCM } = getEncodedCCMAndID(ccm);
+
+		if (!(await this.verifyCCM(context, ccmID))) {
+			return;
+		}
+
+		// create a state snapshot
+		const stateSnapshotID = context.stateStore.createSnapshot();
+		const eventSnapshotID = context.eventQueue.createSnapshot();
+
+		const chainAccountStore = this.stores.get(ChainAccountStore);
+		const beforeCCMForwardingContext: BeforeCCMForwardingContext = { ...context, ccmFailed: false };
+
+		const receivingChainAccount = await chainAccountStore.getOrUndefined(
+			context,
+			ccm.receivingChainID,
+		);
+
+		// do not continue, if
+		// the chain account does not exist
+		// the chain has REGISTERED status
+		if (!receivingChainAccount || receivingChainAccount.status === ChainStatus.REGISTERED) {
+			beforeCCMForwardingContext.ccmFailed = true;
+			if (
```

Comment:
I don't think that this is equivalent to the [current specs](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#forward). The rest of the function should be run anyways.